nvse/nvse/NVSECompiler.cpp

```diff
@@ -0,0 +1,760 @@
+	memcpy(script->data, data.data(), data.size());
+
+	// Debug print local info
+	printFn("[Locals]\n");
+	for (int i = 0; i < script->varList.Count(); i++) {
+		auto item = script->varList.GetNthItem(i);
+		printFn(std::format("{}: {} {}\n", item->idx, item->name.CStr(), usedVars.contains(item->name.CStr()) ? "" : "(unused)"));
+	}
+	printFn("\n");
+
+	// Refs
+	printFn("[Refs]\n");
+	for (int i = 0; i < script->refList.Count(); i++) {
+		const auto ref = script->refList.GetNthItem(i);
+		if (ref->varIdx) {
+			printFn(std::format("{}: (Var {})\n", i, ref->varIdx));
+		} else {
+			printFn(std::format("{}: {}\n", i, ref->form->GetEditorID()));
+		}
+	}
+	printFn("\n");
+
+	// Script data
+	printFn("[Data]\n");
+	for (int i = 0; i < script->info.dataLength; i++) {
+		printFn(std::format("{:02x} ", script->data[i]));
+	}
+	printFn("\n");
+
+	printFn(std::format("\nNum compiled bytes: {}\n", script->info.dataLength));
+
+	return true;
+}
+
+void NVSECompiler::visitNVSEScript(const NVSEScript* nvScript) {
+	// Compile the script name
+	scriptName = nvScript->name.lexeme;
+
+	// Dont allow naming script the same as another form, unless that form is the script itself
+	auto comp = strcmp(scriptName.c_str(), originalScriptName);
+	if (resolveObjectReference(scriptName, false) && comp) {
+		throw std::runtime_error(std::format("Error: Form name '{}' is already in use.\n", scriptName.c_str()));
+	}
+
+	add_u32(0x1D);
+
+	for (auto& global_var : nvScript->globalVars) {
+		global_var->accept(this);
+	}
+
+	for (auto& block : nvScript->blocks) {
+		block->accept(this);
+	}
+}
+
+void NVSECompiler::visitBeginStatement(const BeginStmt* stmt) {
+	// Get opcode
+	auto beginInfo = mpBeginInfo[stmt->name.lexeme];
+
+	// OP_BEGIN
+	add_u16(0x10);
+
+	auto beginPatch = add_u16(0x0);
+	auto beginStart = data.size();
+
+	add_u16(beginInfo.opcode);
+
+	auto blockPatch = add_u32(0x0);
+
+	// Add param shit here?
+	if (stmt->param.has_value()) {
+		auto param = stmt->param.value();
+
+		add_u16(0xFFFF);
+		auto varStart = data.size();
+		auto varPatch = add_u16(0x0);
+
+		if (beginInfo.argType == ArgType::Int) {
+			add_u8('L');
+			add_u32(static_cast<uint32_t>(std::get<double>(param.value)));
+		}
+		else if (beginInfo.argType == ArgType::Ref) {
+			// Try to resolve the global ref
+			auto ref = resolveObjectReference(param.lexeme);
+			if (ref) {
+				auto refInfo = script->refList.GetNthItem(ref - 1);
+				printf("Resolved ref %s: %08x\n", param.lexeme.c_str(), refInfo->form->refID);
+				add_u8('R');
+				add_u16(ref);
+			} else {
+				throw std::runtime_error(std::format("Unable to resolve form '{}'.", param.lexeme));
+			}
+		}
+
+		set_u16(varPatch, data.size() - varStart);
+	}
+
+	set_u16(beginPatch, data.size() - beginStart);
+	auto blockStart = data.size();
+	stmt->block->accept(this);
+
+	// OP_END
+	add_u32(0x11);
+
+	set_u32(blockPatch, data.size() - blockStart);
+}
+
+void NVSECompiler::visitFnDeclStmt(FnDeclStmt* stmt) {
```

Comment:
Inconsistent case strategy, some use snake while others camel, NVSE uses PascalCase. I'm not too bothered about that though